server/src/main/java/de/coronavirus/imis/api/EnumDataController.java

```diff
@@ -0,0 +1,101 @@
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.client.ClientHttpRequestFactory;
+import org.springframework.http.client.ClientHttpResponse;
+import org.springframework.http.client.SimpleClientHttpRequestFactory;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.scheduling.Trigger;
+import org.springframework.scheduling.TriggerContext;
+
+import de.coronavirus.imis.repositories.PatientRepository;
+
+@RestController
+@RequestMapping("/api/enum-data")
+public class EnumDataController {
+  private final PatientRepository patientRepo;
+
+  private static final Pattern RE_TEXTFILE_ENTRY = Pattern.compile("^\\s*(?!#)(?<ENTRY>.*?)\\s*$");
+  private SortedSet<String> predefinedHealthInsuranceCompanies;
+
+  @Autowired
+  public EnumDataController(PatientRepository patientRepo) {
+    this.patientRepo = patientRepo;
+
+    try (BufferedReader hicReader = new BufferedReader(new InputStreamReader(
+      this.getClass().getClassLoader().getResourceAsStream("health_insurance_companies.txt"), "UTF-8"))) {
+
+      this.predefinedHealthInsuranceCompanies = hicReader.lines()
+        .filter((String line) -> RE_TEXTFILE_ENTRY.matcher(line).matches())
+        .map(String::trim)
+        .sorted()
+        .collect(Collectors.toCollection(() -> new TreeSet<>()));
+
+    } catch (IOException ioe) {
+      ioe.printStackTrace();
+      this.predefinedHealthInsuranceCompanies = new TreeSet<>();
+    }
+  }
+
+  @GetMapping("/health-insurance-companies")
+  public HealthInsuranceCompanies getHealthInsuranceCompanies(
+    @RequestParam(required = false) String search,
+    @RequestParam(required = false) Optional<Integer> count) {
+
+    SortedSet<String> userDefined = this.patientRepo.findDistinctInsuranceCompanies();
+    userDefined.removeAll(this.predefinedHealthInsuranceCompanies);
+
+    List<List<String>> sets = Arrays.stream(new SortedSet[] {
+      this.predefinedHealthInsuranceCompanies,
+      userDefined
+    }).map((SortedSet healthInsuranceCompanies) -> {
+      Stream<String> result = ((SortedSet<String>) healthInsuranceCompanies).parallelStream();
+
+      if (search != null) {
+        result = result.filter((var companyName) -> companyName.contains(search));
+      }
+
+      result = result.sorted();
+
+      if (count.isPresent()) {
+        result = result.limit(count.get());
+      }
+
+      return (List<String>) result.collect(Collectors.toList());
+    }).collect(Collectors.toList());
+
+    return new HealthInsuranceCompanies(sets.get(0), sets.get(1));
+  }
```

Comment:
Warum filtern wir im Frontend *und* im Backend?

Reply:
Oder sehe ich richtig, dass das aktuell nicht genutzt wird?

Reply:
**tldr:** Es wird nur im Frontend gefiltert.

**Ausführlicher:**
Das Backend kann die Ergebnisse naiv mittels `String::contains` filtern. In der Evolution des Codes hatte ich zuerst diese serverseitige Filterung genutzt, welche ich allerdings später durch eine verbesserte clientseitige Filterung ersetzt habe. Der Gedanke war den Kommunikationsoverhead zwischen Client und Server, wie auch die Serverlast zu reduzieren, da vorher auf Seite des Clients mit jeder Änderung des Suchfelds eine neue Anfrage an den Server geschickt wurde.

Reply:
Ich finde den Frontendfilter auch geeigneter. Vor allem vor dem Hintergrund, dass jede Anfrage an das Backend mit einer Distinct Search über alle Patienten einhergeht ... defakto getriggert bei jeden getippten Buchstaben. Oder?
Da das ja aber derzeit nicht genutzt wird würde ich vorschlagen es zu entfernen.

Reply:
Jupp, getriggert bei jedem Buchstaben. Die Suchfunktion im Backend schmeiße ich dann jetzt raus, genauso wie den `count` Parameter der durch den Abruf der kompletten Liste sowieso hinfällig ist.

Reply:
Perfekt. Danke!